### ISSUE TITLE
Moving to ES6 Classes

### ISSUE THREAD

**Issue by DefinitelyMaybe** [mrdoob/three.js#19986]
Hi all,

I felt it appropriate to create a new issue (instead of continuing #11552) to further aid everyone in keeping track and update-to-date on the issues and progress surrounding the move to ES6 classes. This should also come in handy for the release doc.

To those wishing to help, look through the list below and let us know what you'd like to work on. [A PR per class is favoured however some folders can be done all at once](https://github.com/mrdoob/three.js/pull/19980#issuecomment-667493107). [If a particular file cannot be converted, make a note at the top of the file](https://github.com/mrdoob/three.js/pull/19934#issuecomment-664241764), or ping me from your PR and I'll note it below.

Notes:

* Keep the `Class.prototype.is**` properties
* class fields are also available if appropriate #20395 (dicussion [here](#20129))
* `new this.contructor()` != `new Foo()` ... [related discussion](https://github.com/mrdoob/three.js/pull/20018#discussion_r472845726).
* Will tick after merged and complete.

## Part 1: src

- [ ] src
  - [x] animation ( #19964, #20014, #20016 , #20013 )
  - [x] audio ( #19975, #20003 )
  - [x] cameras ( #20102, #21623 )
  - [x] core ( #19976, #19977, #19978, #19984, #20008, #21635, #21646 )
  - [x] extras ( #19979, #21624 )
  - [x] geometries ( #19994 )
  - [x] helpers ( #19996 )
  - [x] lights ( #21231, #21232 )
  - [x] loaders ( #19985, #21622 )
  - [x] materials ( #20100, #21626, #21626 )
  - [x] math ( #19980, #19997, #20076, #20089, #21628)
  - [x] objects ( #21266, #21625 )
  - [ ] renderers ( #21053 )
    - [ ] webgl ( #20070 )
    - [x] webxr ( #21646, #21648 )
  - [x] scenes ( #20007 )
  - [x] textures ( #20009 )

## Part 2: examples

- [x] examples
  - [x] animation ( #21596 )
  - [x] cameras ( #21589 )
  - [x] controls (#21629, #21644)
  - [x] curves ( #21593 )
  - [x] effects ( #21610 )
  - [x] environments ( #21585 )
  - [x] exporters ( #21605 )
  - [x] geometries ( #21589 )
  - [x] helpers ( #21583 )
  - [x] interactive ( #21592 )
  - [x] lights ( #21592 )
  - [x] lines ( #21599 )
  - [x] loaders ( #21616, #21614, #21612 )
  - [x] math ( #21598 )
  - [x] misc ( #21618 )
  - [x] modifiers ( #21604 )
  - [x] ~~nodes~~ (won't be changed)
  - [x] objects ( #21600 )
  - [x] postprocessing ( #21621 )
  - [x] renderers ( #21601 )
  - [x] shaders ( #21619 )
  - [x] textures
  - [x] utils ( #21611 )
  - [x] webXR ( #21586 )

**Comment by ianpurvis** [comment]
I'll call dibs on the rest of the audio folder 🔈

**Comment by DefinitelyMaybe** [comment]
hey @mrdoob, could you please clarify how you'd like us to handle the scripts within the examples folder?

I'm liking #19989 going straight for the conversion but I realize in doing so the `utils/modularize.js` can no longer be used on the `examples/js` folder without overwriting that work. Does this mark the end of us maintaining the `examples/js` and the beginning of just `examples/jsm`?

**Comment by yomotsu** [comment]
Could I work on the math folder?

**Comment by DefinitelyMaybe** [comment]
> Could I work on the rest of the math folder?

I'm all for you giving it a shot. I remember trying to a while ago but @Mugen87 ended up telling me off for it, tldr there was test as you go because converting the whole folder in one go might be setting yourself up for a rough ride. A partial, or even file by file, would be welcome I'm sure.

**Comment by ianpurvis** [comment]
@DefinitelyMaybe I'll see what else can be migrated under `src/animation/`

**Comment by linbingquan** [comment]
part of `src/core`: https://github.com/mrdoob/three.js/pull/20008

**Comment by DefinitelyMaybe** [comment]
nice. I think that folder is almost done. Might be just `src/core/Object3D.js` and `src/core/BufferGeometry.js` left?

**Comment by linbingquan** [comment]
> Might be just `src/core/Object3D.js` and `src/core/BufferGeometry.js` left?

Yes,  There are much ES5 class depend on `Object3D` and `BufferGeometry`.

**Comment by linbingquan** [comment]
renderers: https://github.com/mrdoob/three.js/pull/20011

**Comment by DefinitelyMaybe** [comment]
Shoot, there has been some amazing progress on this 🎉

Gonna call dibs on `src/lights`. Opened up the `src/extras` and `src/renderers` in the listing above, turns out there's a couple of folders in each to work through.

**Comment by DefinitelyMaybe** [comment]
Hey all,

what were we doing with regards to this pattern?
```javascript
foo.prototype = Object.assign( Object.create( bar.prototype ), {
	...
	isDirectionalLight: true,
	...
} );
```
is it now:
```javascript
class foo extends bar {
  static isDirectionalLight = true;
  constructor(  ) {
    ...
  }
}
```
or
```javascript
class foo extends bar {
  constructor(  ) {
    this.isDirectionalLight = true;
  }
}
```
currently, I've been doing the second one but looking at it, I'd say the intention of this might've been closer to the first.

**Comment by ianpurvis** [comment]
> ```js
> class foo extends bar {
>   constructor(  ) {
>     this.isDirectionalLight = true;
>   }
> }
> ```

☝️ This is right. `isDirectionalLight` is a good example, it gets used like this:

```sh
$ git grep 'isDirectionalLight\b' src/ examples/js ':!*.ts'
examples/js/exporters/GLTFExporter.js:                  if ( light.isDirectionalLight ) {
examples/js/exporters/GLTFExporter.js:                  } else if ( object.isDirectionalLight || object.isPointLight || object.isSpotLight ) {
examples/js/renderers/SVGRenderer.js:                   } else if ( light.isDirectionalLight ) {
examples/js/renderers/SVGRenderer.js:                   if ( light.isDirectionalLight ) {
src/lights/DirectionalLight.js: isDirectionalLight: true,
src/renderers/webgl/WebGLLights.js:                     } else if ( light.isDirectionalLight ) {
```

That said, there could be performance gains by retaining some prototype properties...

```js
class foo extends bar {
}
foo.prototype.baz = heavyThing;
```

Maybe we can review those in the PR on a case by case basis.

**Comment by ianpurvis** [comment]
I'll take a shot at  `src/renderers/webgl` 👍 👍

**Comment by DefinitelyMaybe** [comment]
good luck, have fun. that one's got a fair bit going on in it

**Comment by linbingquan** [comment]
webxr: https://github.com/mrdoob/three.js/pull/20038

**Comment by ianpurvis** [comment]
Hey all- I've got `src/renderers/webgl` converted. It was crazy.  I'll keep reviewing everything and wait for #20039 before I start pushing the PRs.

**Comment by DefinitelyMaybe** [comment]
👍 
looking forward to it

**Comment by ianpurvis** [comment]
I noticed @yomotsu used read-only getters for `is*` properties in the math PR... Maybe that's the best!

```js
class foo extends bar {
  get isDirectionalLight() {
    return true;
  }
}
```

**Comment by DefinitelyMaybe** [comment]
hmm it might be. I know others have tried in places to make variables unchangeable. This looks like a pretty decent pattern to me.

[quick jsfiddle show case](https://jsfiddle.net/8fahvg3o/)

**Comment by mrdoob** [comment]
Seems like this is the way to go:

```js
foo.prototype.isDirectionalLight = true;
```

**Comment by DefinitelyMaybe** [comment]
going to look into `src/objects`. will see how far I can get without breaking things.

**Comment by yomotsu** [comment]
The math folder was just completed.
([Interpolates classes are excluded.](https://github.com/mrdoob/three.js/pull/20086#issuecomment-674514866))

**Comment by DefinitelyMaybe** [comment]
hey all,
I'm going to pull back on working through `src/objects`, someone else is more than welcome to pick it up in the mean time.

**Comment by marcofugaro** [comment]
Hey, good job with this guys! 💪

Just one issue, `foo.prototype.isDirectionalLight = true;` doesn't allow for tree-shaking still, since the class is modified in place.
Also at this point I would strongly suggest not to touch the prototype anymore since we're using classes now and the prototype can be abstracted.

What about this pattern? It looks the most explicit to me.

```js
constructor( x = 0, y = 0, z = 0 ) {

  Object.defineProperty( this, 'isVector3', {
    value: true,
    enumerable: false,
    writable: false,
    configurable: false,
  } );
```

There is also the short version, which does the same thing but is a bit more implicit.

```js
constructor( x = 0, y = 0, z = 0 ) {

  Object.defineProperty( this, 'isVector3', { value: true } );
```

**Comment by mrdoob** [comment]
@marcofugaro the short version looks good to me 👍

**Comment by marcofugaro** [comment]
Ok, did PR.

**Comment by marcofugaro** [comment]
Also, I just noticed we can't use class fields because we're blocked by [this PR](https://github.com/bublejs/buble/pull/169).

I can update the build pipeline to use the more popular [@rollup/plugin-babel
](https://github.com/rollup/plugins/tree/master/packages/babel) and fix that issue if you guys want.

**Comment by mrdoob** [comment]
> I can update the build pipeline to use the more popular @rollup/plugin-babel
and fix that issue if you guys want.

Yeah, it would be great if you could do a PR 🙏

**Comment by DefinitelyMaybe** [comment]
@marcofugaro , @mrdoob yay! yes please

**Comment by ianpurvis** [comment]
Hey everybody, quick heads-up that I've updated #20014 with `Object.defineProperty` and also fixed the subclassing of `AnimationClip`. If anyone has time to review it, I would appreciate the extra 👀  🙇 🙇

**Comment by mrdoob** [comment]
Okay. How that `r120` is out... Could someone check that things are being tree-shaked properly?

**Comment by yomotsu** [comment]
Seems it kinda works for me, but slightly.
I don't see `ArrowHelper` anymore in my bundle file with webpack. but many unnecessary code are still exist😢

**Comment by DefinitelyMaybe** [comment]
how about bundle size? before and after

**Comment by ianpurvis** [comment]
Hey everybody, I had an idea and spiked a tool called [shakediff](https://github.com/ianpurvis/shakediff). You can run it like this:

```sh
$ shakediff builds/three.module.js Color
```

That will generate a small module importing `Color`, bundle it with three, and then create a diff of three from rollup's metadata. You can either work with the detailed diff or just pipe it to `diffstat` to get the high level:

```sh
$ shakediff build/three.module.js Color | diffstat 
 three.module.js | 2878 --------------------------------------------------------
 1 file changed, 1 insertion(+), 2877 deletions(-)
```

Right now there's no package, but you can install it from my repo if you want to try it out. Feedback appreciated!  ✌️ 
```sh
npm -g i ianpurvis/shakediff.git
```

**Comment by arodic** [comment]
> hey @mrdoob, could you please clarify how you'd like us to handle the scripts within the examples folder?
> I'm liking #19989 going straight for the conversion but I realize in doing so the utils/modularize.js can no longer be used on the examples/js folder without overwriting that work. Does this mark the end of us maintaining the examples/js and the beginning of just examples/jsm?

What is the verdict on this one @mrdoob? Would it make sense to make jsm the source of truth so we can convert examples to ES6 classes? Perhaps we can made utils/demodularize.js to support conversion in the other direction?

**Comment by donmccurdy** [comment]
The `examples/js` folder will be removed in December 2020. Until then, we should leave the folder as it is, without upgrading its contents — or the `examples/jsm` files generated from its contents — to ES6 Classes. After that date, The `examples/jsm` files will become the source of truth, and can be updated to ES6 Classes.

**Comment by ianpurvis** [comment]
I noticed that module-scope variables are not being removed during shaking. You can see the behavior with these four vars, `build/three.module.js:43059`

```js
const _position$3 = new Vector3();
const _quaternion$4 = new Quaternion();
const _scale$2 = new Vector3();
const _orientation$1 = new Vector3();
```

Looks like we should mark these pure with the special comment-

```js
const _position = /*@__PURE__*/ new Vector3();
```

Does that work for everybody?

**Comment by mrdoob** [comment]
Sounds good to me 👌

**Comment by ianpurvis** [comment]
Nice, that deletes another 144 lines when tree shaking with `Color` 🥳 

Please everyone give your open PRs the `/*@__PURE__*/` treatment when you have time!

**Comment by marcofugaro** [comment]
@ianpurvis great! are you using rollup, right? could you share your config?

@mrdoob I finally got around to do some testing. I personally use webpack, so I did some tests with it, [my config is this if anyone is interested](https://github.com/marcofugaro/threejs-modern-app/blob/master/webpack.config.js).

I tested this code

```js
import * as THREE from 'three'

console.log(THREE.WebGLRenderer)
```

### 0.119.1
<img width="323" alt="0 119 1" src="https://user-images.githubusercontent.com/7217420/91854038-6514c980-ec63-11ea-9a20-d4d739044246.png">

### 0.120.1
<img width="327" alt="0 120 1" src="https://user-images.githubusercontent.com/7217420/91854067-72ca4f00-ec63-11ea-9246-abf80529b4c9.png">

So something is being tree-shaked, great! 

Looking into it more, it looks like the class `AudioListener` isn't in the bundle, great news! 

<img width="350" alt="Screenshot 2020-09-01 at 16 49 22" src="https://user-images.githubusercontent.com/7217420/91869350-115fab80-ec76-11ea-9688-2d7e29f781d3.png">

<img width="450" alt="Screenshot 2020-09-01 at 16 50 07" src="https://user-images.githubusercontent.com/7217420/91869372-1886b980-ec76-11ea-9754-31f371ffe591.png">

Webpack also automatically removes the unused variables pointed out by @ianpurvis.

------------------------------

After this, I decided to test static methods defined outside the class

<img width="350" alt="Screenshot 2020-09-01 at 16 50 37" src="https://user-images.githubusercontent.com/7217420/91869897-b2e6fd00-ec76-11ea-8e50-a24fcdda1311.png">

 and unfortunately, **this makes the class non tree-shakeable**

<img width="450" alt="Screenshot 2020-09-01 at 16 51 04" src="https://user-images.githubusercontent.com/7217420/91869911-b7131a80-ec76-11ea-96d9-0608317effb3.png">

------------------------------

After doing some more digging, I noticed geometry classes like `DodecahedronGeometry`, which weren't used anywhere, were still in the bundle

<img width="350" alt="Screenshot 2020-09-01 at 16 51 29" src="https://user-images.githubusercontent.com/7217420/91870730-559f7b80-ec77-11ea-93bd-df0f1b9785ba.png">

<img width="450" alt="Screenshot 2020-09-01 at 16 52 08" src="https://user-images.githubusercontent.com/7217420/91870804-5a642f80-ec77-11ea-9d9e-9b2b19d3def6.png">

Later I discovered, this was due to this `Geometries` object in the `three.module.js` bundle 

<img width="500" alt="Screenshot 2020-09-01 at 17 18 32" src="https://user-images.githubusercontent.com/7217420/91871585-8da6be80-ec77-11ea-9ac4-ac825324d78e.png">

Which is generated automatically when using [patterns like this in the ObjectLoader](https://github.com/mrdoob/three.js/blob/3f1cf4e8c5ce4db04ca61283c8b0a3fc1d526537/src/loaders/ObjectLoader.js#L183-L188). **So that pattern will need to change if we want to make `src/geometries` tree-shakeable.**

**Comment by donmccurdy** [comment]
> Which is generated automatically when using patterns like this in the ObjectLoader. This will probably go away when we make ObjectLoader a class, and src/geometries will be tree-shakeable.

ObjectLoader really does depend on all of those things, and presumably it isn't tree-shakeable because it's not an ES6 Class yet. Maybe converting it would fix that?

**Comment by mrdoob** [comment]
@marcofugaro https://github.com/mrdoob/three.js/pull/20239

**Comment by ianpurvis** [comment]
@marcofugaro No problem, here's the rollup [config](https://github.com/ianpurvis/shakediff/blob/b860cdb1a8f1e7758d2608ca14558cc27f137731/shakediff.js#L25-L38)

**Comment by ianpurvis** [comment]
@marcofugaro Hi, I added webpack support to shakediff... does this [webpack config](https://github.com/ianpurvis/shakediff/blob/trunk/src/pack.js#L16-L73) seem ok to you? Since terser is responsible for dead code elimination in webpack's tree-shaking, it is difficult to prevent some transformation artifacts in those outputs. But if you diff using a histogram algorithm with whitespace disabled it is manageable. From a quick test tonight, webpack seems to accept both `/*@__PURE__*/` and `/*#__PURE__*/`. I think we can standardize on one or the other. More tomorrow...

**Comment by mrdoob** [comment]
@ianpurvis I don't think I'm allowed to install `shakediff` on my machine...

<img width="839" alt="Screen Shot 2020-09-22 at 9 59 04 AM" src="https://user-images.githubusercontent.com/97088/93862855-a85fd680-fcfd-11ea-9ea8-89daf14b365a.png">

**Comment by mrdoob** [comment]
Anyway, I did a simple test:

```js
import { WebGLRenderer } from './src/renderers/WebGLRenderer.js';
console.log( new THREE.WebGLRenderer() );
```
And noticed that `Geometry` was not being tree-shacked. https://github.com/mrdoob/three.js/pull/20394 fixes it.

**Comment by mrdoob** [comment]
I then tried doing:

```js
import { Vector } from './src/math/Vector3.js';
console.log( new Vector3() );
```

And noticed that rollup doesn't tree-shake unused methods... 😕

**Comment by marcofugaro** [comment]
@mrdoob unfortunately class methods are never gonna be tree-shakeable by any tool. Heck, they can't even be minified!

This is because in javascript, you can access methods like `this['I am a string']`, just like you can access dynamically object properties. For this reason, tools don't know in advance if and how you're gonna call a method (could be `this[variable]`).

This is the same for both ES6 Classes and Function classes. For example, in `three.min.js`, the methods are left untouched, just like any object property 🙂

**Comment by mrdoob** [comment]
And no one has proposed a builder "strict" mode, where you're expected not to call methods like `this['I am a string']`?

**Comment by marcofugaro** [comment]
No idea 🤷‍♂️

Here is a discussion about this topic regarding rollup: https://github.com/rollup/rollup/issues/349

**Comment by donmccurdy** [comment]
> class methods are never gonna be tree-shakeable by any tool. Heck, they can't even be minified!

Sorry for going off topic, but you can get minification to work if you give the compiler some hint as to what's "public" vs. "private". In the past I've used "_" prefixes on methods for this. See https://github.com/developit/microbundle/wiki/mangle.json.  But, unfortunately, I also have no idea whether anything similar is possible for tree-shaking. 😕

**Comment by DefinitelyMaybe** [comment]
Wooh! #20395 merged! good stuff @marcofugaro

**Comment by ianpurvis** [comment]
Awesome news about babel and geometry!

> @ianpurvis I don't think I'm allowed to install shakediff on my machine...

@mrdoob Hm, looks like Parcel has a dependency on that version of fsevents... maybe I can pin that to something higher.

**Comment by ianpurvis** [comment]
Fyi, good info here about [tree-shaking improvements](https://github.com/webpack/changelog-v5#major-changes-optimization) in Webpack 5...

**Comment by mrdoob** [comment]
> Fyi, good info here about [tree-shaking improvements](https://github.com/webpack/changelog-v5#major-changes-optimization) in Webpack 5...

Progress...

**Comment by devingfx** [comment]
Hi all,

What are the goals of `.is**Classname**` property and `this.type = **Classname**` exactly?
Isn't it a rest of old pseudo class pattern?
Why don't get rid totally of it and replace this usage by the normal type checking way?
```javascript
obj instanceof Vector3
// or
obj.constructor === Vector3
```
I mean, as part of moving to ES classes, objects will have correct types already checkable...

**Comment by mrdoob** [comment]
The library used `obj instanceof Vector3` before.
But Rich Harris implemented and changed all the checks to `is*` to allow tree-shaking.

**Comment by devingfx** [comment]
> The library used `obj instanceof Vector3` before.
> But Rich Harris implemented and changed all the checks to `is*` to allow tree-shaking: #9310

Thanks for the answer, I had read the thread, btw I don't understand why a class have to be avoided to be in the output if some other class needs to type-check it... I mean if, bundled class X needs to be aware that an object is a class B, then class B is used (at least to create the said object somehow) and have to be bundled, isn't it?

**Comment by donmccurdy** [comment]
For example, WebGLRenderer needs to check whether a geometry it renders is BufferGeometry or Geometry, but it never creates Geometry instances. Most three.js applications should only use BufferGeometry, and so we want tree-shaking to eliminate Geometry (and its subclasses) from the bundle when possible.

**Comment by mrdoob** [comment]
In a visual way...

This pattern makes `WebGLRenderer` always include `Geometry` when bundling:

```js
import { Geometry } from '../core/Geometry.js';
import { BufferGeometry } from '../core/BufferGeometry.js';

if ( geometry instanceof Geometry ) {
} else if ( geometry instanceof BufferGeometry ) {
}
```

This pattern does not:

```js
if ( geometry.isGeometry === true ) {
} else if ( geometry.isBufferGeometry === true ) {
}
```

**Comment by DefinitelyMaybe** [comment]
hey all,

[given this](https://github.com/mrdoob/three.js/issues/20455#issuecomment-706954770) what's our thoughts on moving forward from here?

**Comment by donmccurdy** [comment]
I don't think there's any change, yet. Anything that (1) isn't in `examples/js` and (2) isn't extended by something in `examples/js` can be converted to ES6 Classes. If that process is wrapping up, we should decide when to begin switching `examples/js` to classes.

**Comment by DefinitelyMaybe** [comment]
From memory, and after a quick look we've done a fair bit and there's still some PR's ready in waiting.

I might try re-compile my dependencies/extended-in-examples list.

**Comment by ianpurvis** [comment]
@DefinitelyMaybe Hey! That sounds great, your list was super helpful. And @donmccurdy's strategy makes sense to me. 👍 I think it would be best to get closure on the work we've already done. #20070 would be a great PR to attack because it can give strategy for migrating private/hidden vars to ES6 (we need this for migration of `renderers/webgl` to ES6). If everybody can run the benchmarks or chime in to the discussion there, I would appreciate it. I think we have some decent options, just need to make sure performance is good.

**Comment by devingfx** [comment]
Okay, 
I may start some PR for examples rewrite to ES classes...
@mrdoob is it a problem for you doing the jsm to js script or not? (I think not but tell me if you prefer us to wait it)

**Comment by Mugen87** [comment]
Before converting the example code to classes, #20527, #20529 or a different solution needs to be merged first.

**Comment by devingfx** [comment]
@DefinitelyMaybe You said in [1st comment](https://github.com/mrdoob/three.js/issues/19986#issue-671461599) :

> - use class fields

1) Does it means we can use class fields for everything ?
Like so :

```javascript
class Light extends Object3D {
	
	type = 'Light';
	isLight = true;

	color = new Color;
	intensity = 1;
	
	constructor(color, intensity = 1) {

		super();

		this.color = new Color(color);
		this.intensity = intensity;

	}

	copy() {
		...
	}
}
```

... or it's for `.is*` property only ?

So then, 2) what about empty constructors? ES2015 spec said constructors with only a `super()` call is implicit if not provided, so some child classes can be really defined easier :

```javascript
class MapControls extends OrbitControls {

	screenSpacePanning = false; // pan orthogonal to world-space direction camera.up

	mouseButtons = { LEFT: MOUSE.PAN, MIDDLE: MOUSE.DOLLY, RIGHT: MOUSE.ROTATE };

	touches = { ONE: TOUCH.PAN, TWO: TOUCH.DOLLY_ROTATE };

}
```

And 3) what about private class fields ?

```javascript
class OrbitControls extends EventDispatcher {
	...

	#offset = new Vector3();

	// so camera.up is the orbit axis
	#quat = new Quaternion().setFromUnitVectors( object.up, new Vector3( 0, 1, 0 ) );
	#quatInverse = quat.clone().inverse();

	#lastPosition = new Vector3();
	#lastQuaternion = new Quaternion();

	#twoPI = 2 * Math.PI;
	
	update() {
		... ( no more closure and return function here )
	}

}
```

**Comment by DefinitelyMaybe** [comment]
#20395

The key is to pass all the tests.

@marcofugaro I believe there are still some variables that can be changed into class fields, right?
I'm looking at things like:

```javascript
class EdgesGeometry extends BufferGeometry {

	constructor( geometry, thresholdAngle ) {

		super();

		this.type = 'EdgesGeometry';
```

changed to:

```javascript
class EdgesGeometry extends BufferGeometry {

	type = 'EdgesGeometry';

	constructor( geometry, thresholdAngle ) {

		super();
```

It is likely that we'll run into some other caveats around the place, like how I ran into "`new this.contructor() != new Foo()`".

> private class fields

this is an on-going discussion. [It'll be a while before we use it if we use it.](https://caniuse.com/mdn-javascript_classes_private_class_fields) I'm not sure of a issue or PR that I can point you towards.

**Comment by marcofugaro** [comment]
> @marcofugaro I believe there are still some variables that can be changed into class fields, right?
I'm looking at things like:

Yeah that can be done now. However we can't do that for `.is*` properties since they have to be non-enumerable and non-writable, for those we have to use `Object.defineProperty(this, ...`.

**Comment by DefinitelyMaybe** [comment]
how is it checked again? would it be possible for us to use the `static` keyword instead of `Object.defineProperty(this, ...`?

**Comment by devingfx** [comment]
> how is it checked again? would it be possible for us to use the `static` keyword instead of `Object.defineProperty(this, ...`?

I don't think so, because `obj.is*` have to be on instances, not on the class itself...

I'm not sure what does babel transpile exactly in current config, but we may use decorators to set a class field as non-enumerable/non-writtable:
```javascript
import { unwrittable, unenumerable } from 'some/decorator/helpers.js'

class Foo {
    @unwrittable@unenumerable
    isFoo = true;
}
```

**Comment by marcofugaro** [comment]
@DefinitelyMaybe `static` properties are different, they aren't accessible from the instance, rather from the class itself.

```js
class Test {
  static staticProp = true

  constructor() {
    Object.defineProperty(this, 'definedProp', { value: true })
  }
}
const test = new Test()

console.log(test.staticProp, test.definedProp)
```
result: 
```js
undefined true
```

**Comment by DefinitelyMaybe** [comment]
yes, yes.

My main point being that if we were to adjust the code that checks this property we could go from:
```javascript
class Test {
  constructor() {
    Object.defineProperty(this, 'definedProp', { value: true })
  }
}
```
to
```javascript
class Test {
  static definedProp = true
  constructor() {
  }
}
```

So I was wondering where and why this check takes place and whether or not we could change it.

**Comment by devingfx** [comment]
@DefinitelyMaybe The problem is getting the information of an instance type. So if you could access the class of the instance to get its static prop, then why having a static prop? You already have the name of the class...
```javascript
obj.constructor.isFoo == true
obj.constructor.name == 'Foo'
```

Edit: writing this I could figure out that having several .is* can be useful to test on each inheritance chain finally, compared to only the single final class name...
```JavaScript
obj.constructor.isFoo || obj.constructor.isBar || obj.constructor.isBaz
```

**Comment by DefinitelyMaybe** [comment]
> from the instance

oops. re-read that one. yup, never mind what I was saying.

```javascript
if ( scope.object.isPerspectiveCamera ) {
...
}
```

```javascript
if ( geometry.isBufferGeometry ) {
...
}
```

```javascript
if ( material.isShaderMaterial ) {
...
}
```

**Comment by DefinitelyMaybe** [comment]
right. Two things.
 - What have we got left?

> we should decide when to begin switching `examples/js` to classes.

 - Getting closer to needing to discuss this in order to progress.

I don't know whether we should wait for #20527 or #20529 because both of them will be trying to re-create the `examples/js` in it's current form, which is not where we're aiming for. My initial suggestion is to start converting `examples/js` as is. maybe something like:

```javascript
THREE.CinematicCamera = function ( fov, aspect, near, far ) {
...
}
```

to

```javascript
THREE.CinematicCamera = class {
  constructor( fov, aspect, near, far ){
  ...
  }
...
}
```

**Comment by DefinitelyMaybe** [comment]
I also wanted to re-iterate something that @mrdoob [said recently](https://github.com/mrdoob/three.js/issues/20455#issuecomment-706954770) 

> I would hate forcing newbies to learn about polyfills or bundlers in order to render their first cube.

As someone who's been working with modules for the last wee while, the workflow is clear and I understand the various concepts needed to get something working. We don't need bundlers and polyfills but we will need to adjust how Three.js is initially [described](https://threejs.org/docs/index.html#manual/en/introduction/Creating-a-scene).

**Comment by devingfx** [comment]
We will need also to clarify how to replace some non-class patterns like:

```javascript
function Foo() {

	this.update = function () {
	
		var priv = 'foo'

		return function update() {
			...
		};

	}();

}
```

IMHO we could use private class fields, and make the rollup/babel transpile to old pattern for certain old browser don't implementing this natively...

**Comment by marcofugaro** [comment]
> IMHO we could use private class fields, and make the rollup/babel transpile to old pattern for certain old browser don't implementing this natively...

I agree with this strategy, but the decision of course depends on the maintainers that will need to maintain that code.

**Comment by DefinitelyMaybe** [comment]
cool. It might be an idea to show this off within `src` first, i.e. find a similarly strange pattern to what @devingfx described within `src`, make a PR that uses private variables instead and show what babel does with it.

**Comment by DefinitelyMaybe** [comment]
searches: `internal`, `private` 👀

... there are many options.

**Comment by ianpurvis** [comment]
How about WebGLAnimation, it's a nice small class... https://github.com/mrdoob/three.js/pull/20070

**Comment by devingfx** [comment]
> any suggestions on which script?

I'm focused on `exemples/js` and I found this on OrbitControls ...

> show what babel does with it.

I'm pretty sure it won't output something matching mrDoob's syntax thought 😅

**Comment by DefinitelyMaybe** [comment]
to right.

and oops. Where is my brain at... we will need one of #20527 or #20529 (or other) to merge in order to begin transforming the `examples` folder. `examples/js` is sticking around for the foreseeable future which means touching the `examples/js` folder with classes is an absolute no no. that'd break compatibility with IE... sigh.

> make the rollup/babel transpile to old pattern

Guess its thinking through and supporting those JSM to JS PR's. We can take aim at the `examples/jsm` once a decision is reached.

Adding private variables into `src` is still a good idea as it would be good solution to the long standing issue of "please don't play with those variables, they've got a specific purpose"

**Comment by devingfx** [comment]
> touching the `examples/js` folder with classes is an absolute no no. 

Ooops, my bad, I was talking about digging into `exemples/jsm` of course, `exemples/js` will be the "built" version in near future...

> that'd break compatibility with IE... sigh.

What??? Are we still talking about Internet Explorer in 2020 ? Are we talking about the compatibility of a WebGL library on this 7yr old dinosaur? Seriously !! [1.4%](https://www.caniuse.com/es6-class) ...
_(we should add a statistics collecter in the lib to gather if any THREE user is planing a use under IE)_

😞 BTW, it's why babel exists...

**Comment by Mugen87** [comment]
The build files `three.js` and `three.min.js` as well as all files in `examples/js` should still support older browsers like IE 11.

There was a PR last year that should stop support for IE 11 (#18091) but it turned out that there are still users relying on this browser. And the current policy of the project is to provide respective files for the user so they don't to perform the ES6 to ES5 conversion by themselves. This was also discussed in #20455.

**Comment by devingfx** [comment]
>  the current policy of the project is to provide respective files for the user so they don't to perform the ES6 to ES5 conversion by themselves.

Okay, no problems with that policy if the source can be developed a modern way, and if resulting builds are not supposed to be readable but just working... 
Because transpilers outputs such ugly code !
So I see no problems for `src` being written in ESnext, and builds being ugly, BUT it's a potential problem with `exemples/js` if this files should be readable, commented and well formed...

BTW I'm asking myself for several time why some exemples are exemples and not main core sources !? 
Exemple: controls are usualy used as is, copied like optional addons and not used as a real exemple you read and draw inspiration from to write your projects, like a rotating cube exemple for instance...
_(I started my journey with THREE by a search: "webgl rotating cube" and an exemple of it leading to a 1 night code marathon developing a small game with a cube moving on platforms ^^)_

**Comment by Mugen87** [comment]
When using a proper build then it does not matter what files are in the core and what in the examples directory. As long as tree shaking properly works you only have the source files in your build which are actually required.

To me, the distinction of core and examples comes from a time where this kind of workflow was not yet available. The core should be small and compact since you had to import it completely as a single component. Only the most _important_ files should be included. What files ended up in the core was in some sense a case-by-case decision.

**Comment by devingfx** [comment]
> When using a proper build then it does not matter what files are in the core and what in the examples directory. As long as tree shaking properly works you only have the source files in your build which are actually required.

This is true for ES modules users only ;)

**Comment by Filyus** [comment]
There also Three.Legacy.js, which contains some dirty Object.assign code for modules in other files.
I think such code must be located inside the corresponding module files.

**Comment by Mugen87** [comment]
The idea of having a separate file `Three.Legacy.js` was that custom builds can easily remove it and thus achieve smaller build sizes. It will be more hard to achieve this if the legacy code is integrated back to the class files.

**Comment by Filyus** [comment]
It seems to me that most of the legacy code will be removed after Tree-Shaking.

**Comment by Mugen87** [comment]
When a custom build is used that bundles the entire application + engine then yes. Assuming the build tool supports tree shaking.

However it won't work for users who fork the repo and build a custom version of `three.js` or `three.min.js` with the build scripts of the repository.

**Comment by Filyus** [comment]
I just want to say that the tree-shakable npm module would be perfect.
It might make sense to completely delete this file in this case.

**Comment by Mugen87** [comment]
Related #13537.

Maybe we could consider to delete old stuff (e.g. older than 2 years) and move the remaining code to the modules.

**Comment by mrdoob** [comment]
We could also remove `Three.Legacy.js` from `three.module.js` and make a new build `three.compatible.module.js`?

**Comment by Filyus** [comment]
It sounds great.

**Comment by mrdoob** [comment]
While converting [ui.js to ES6 classes](https://github.com/mrdoob/three.js/pull/21229) I realised this works:

```js
class Foo {
    constructor(){
        this.bar = 1;
    }
}

Foo.prototype.isFoo = true;
```

Can someone remind me whey we decided to do this in the constructor instead?

```js
Object.defineProperty( this, 'isFoo', { value: true } )
```

**Comment by mrdoob** [comment]
Ah, just re-read @marcofugaro comment:

> Just one issue, `foo.prototype.isDirectionalLight = true;` doesn't allow for tree-shaking still, since the class is modified in place.

**Comment by marcofugaro** [comment]
Yup, also with `defineProperty` the property is non-writable.

**Comment by mrdoob** [comment]
This seems to work too:

```js
class Foo {
    constructor() {
        this.bar = 1;
    }
    isFoo = true
}
```

Is there a downside to this?

**Comment by mrdoob** [comment]
Hmm, it seems like Babel converts it to `this.isFoo = 1` inside the constructor.

**Comment by marcofugaro** [comment]
> Hmm, it seems like Babel converts it to this.isFoo = 1; inside the constructor 🤷‍♂️

Yup, that's basically what a class property is.

> Is there a downside to this approach?

When declaring the `isFoo` property like that,

```js
class Foo {
    isFoo = true
    constructor() {
      // ...
    }
}
```

it becomes `writable`, meaning an end user could potentially modify it (but we might be fine with this since this scenario is highly improbable):
```js
const foo = new Foo();
foo.isFoo = false; 

console.log(foo.isFoo); // --> false
```

It also becomes `enumerable`, meaning it will show up in the keys of the object when you loop them:
```js
const vector = new Vector3();
for (let key in vector) {
  console.log(key); // --> x, y, z, isVector3
}
```
and this was the reason because we preferred `defineProperty`, because these kind of unit tests would fail otherwise:

https://github.com/mrdoob/three.js/blob/fade4fa03bae7da0cd9bc2f459fa96ee45568303/test/unit/src/core/Object3D.tests.js#L300-L304

**Comment by donmccurdy** [comment]
> We could also remove Three.Legacy.js from three.module.js and make a new build three.compatible.module.js?

We should consider it an eventual goal to be able to either (a) set `sideEffects: false` in `package.json`, or (b) have only a short list of files included there. See [Tree Shaking • Marking the file as side effect free](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free).

As far as I know, `Three.Legacy.js` is the main obstacle to setting `sideEffects: false` now. If we omit it from `three.module.js` that would let us set `sideEffects: false`, but it also means that users cannot really get the compatibility warnings and use any part of `examples/jsm` at the same time.

Another option would be to move the legacy warnings to their own file, in `examples/jsm` or elsewhere, and then mark that file as having side effects. Something like `sideEffects: ["examples/jsm/legacy.js"]`. Users can simply import this file and it will "patch" all of the other three.js objects, if I understand correctly:

```js
import { Scene } from 'three';
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js';
import 'three/examples/jsm/legacy.js'; // Optional, side-effects only import.
```

**Comment by mrdoob** [comment]
Wouldn't removing `Three.Legacy.js` from `three.module.js` allow use to se `sideEffects: false`?
Or does it affect `build/three.js` and `build/three.min.js` too?

**Comment by donmccurdy** [comment]
I believe that's correct and safe. `three.js` and `three.min.js` are not modules and will never be tree-shaken, so `sideEffects: false` should have no effect on them. The remaining question then is just how to make `Three.Legacy.js` warnings available to those using modules.

**Comment by marcofugaro** [comment]
> The remaining question then is just how to make Three.Legacy.js warnings available to those using modules.

What about just moving each `Three.Legacy.js` method into their respective classes?

**Comment by donmccurdy** [comment]
That's an option too. I'm assuming it was originally a separate file because maintainers found it cleaner that way, not sure if that's a strong preference. The upside to putting it elsewhere would be that it's another 7.5kb that users of ES modules can tree-shake when they don't need it.

**Comment by mrdoob** [comment]
> The remaining question then is just how to make `Three.Legacy.js` warnings available to those using modules.

Importing `three/build/three.compatible.module.js` wouldn't work?

>> What about just moving each `Three.Legacy.js` method into their respective classes?
>
> That's an option too. I'm assuming it was originally a separate file because maintainers found it cleaner that way

Correct. It was horrible.

**Comment by donmccurdy** [comment]
> Importing three/build/three.compatible.module.js wouldn't work?

I don't know, this might be a question for someone more familiar with bundler tools. I'm not sure if `sideEffects: false` will apply to the example below, but if it did, the legacy warnings would get tree-shaken and never used. The other downside is that this will import two copies of the library:

```js
import { Scene } from 'three/build/three.compatible.module.js';
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js';
```

**Comment by ianpurvis** [comment]
What do people think of the PIXI.js approach? We could wrap all of `Three.Legacy.js` with a function... 

https://github.com/mrdoob/three.js/pull/20240#issuecomment-688648978

> I also like [PIXI's approach](https://pixijs.download/v5.3.3/docs/PIXI.html#.useDeprecated) to deprecations where we turn Three.legacy.js into a single exported function. Something like this-
> 
> import * as THREE from 'three';
> THREE.enableDeprecations(); // MUST be bound to namespace
> Since the function itself is tree-shakable, this allows elimination of all the legacy code in one shot, regardless of which classes are imported by the user. And it touches less files. 😉

**Comment by donmccurdy** [comment]
Oh! That sounds much easier, yes. 

I don't understand why it would need to be bound to the namespace though, seems like it should work either way?

**Comment by ianpurvis** [comment]
I think the namespace binding just allows injection of classes which are not otherwise exported

```js
import * as THREE from 'three'
console.log(THREE.DeprecatedClass) // -> undefined
THREE.enableDeprecations()
console.log(THREE.DeprecatedClass) // -> defined
```

Might be something that you can go off-road with, though

```js
import { Scene, OrbitControls, enableDeprecations } from 'three'
enableDeprecations.call({ Scene, OrbitControls })
```

**Comment by donmccurdy** [comment]
Hadn't thought of entire classes being deprecated, that makes sense, but it only works if the user is also importing `* as THREE` anyway, and if they're doing that they've lost tree-shaking... I think those deprecated classes should just continue to be exported normally, and `enableDeprecations()` would just modify prototypes of non-deleted classes rather than attaching things to the namespace.

^But this is getting into the weeds maybe. I do like the idea of moving deprecations into an exported function better than trying to make a new `build/three.legacy.module.js` output or `examples/jsm/legacy.js` file.

**Comment by ianpurvis** [comment]
Yeah, I think it could simplify maintenance. Maybe a little clunky architecturally tho... Going in the other direction, to get a nice import scheme that tree shakes well, I think you indeed might have to build & distribute something like `three.legacy.module.js`. For the dirty prototype issue, maybe `src/Three.legacy.js` could get refactored with import aliasing-

```js
import { Sphere as _Sphere } from './math/Sphere.js';
export const Sphere = (function () {
	Object.assign( _Sphere.prototype, {
		empty: function () {
			console.warn( 'THREE.Sphere: .empty() has been renamed to .isEmpty().' );
			return this.isEmpty();
		}
	} ); 
	return _Sphere
})();
```

Might be a little cleaner with es6 inheritance
```js
import { Sphere as _Sphere } from './math/Sphere.js';
export class Sphere extends _Sphere {
	empty() {
		console.warn( 'THREE.Sphere: .empty() has been renamed to .isEmpty().' );
		return this.isEmpty();
	}
}
```

Then people could do this
```js
import { Mesh } from 'three'
import { Sphere } from 'three/legacy'
```

Should also be possible to add enough `export { foo } from 'bar'` statements to `src/Three.legacy.js` that it becomes a superset of `three.module.js` where people could do this

```js
import { Mesh, Sphere } from 'three/legacy'
```

Anyways... just some thoughts! ✌️

**Comment by donmccurdy** [comment]
I was imagining just reorganizing `src/Three.legacy.js`, and leaving it in `build/three.module.js`, with `sideEffects: false` set. Could we structure it as:

**src/Three.legacy.js**

```js
import { Mesh } from './objects/Mesh.js'
import { Points } from './objects/Points.js';
// ...

export class DeprecatedClass1 { ... }

export class DeprecatedClass2 { ... }

export function useDeprecatedMethods () {

  Object.assign( Mesh.prototype, {
    deprecatedMethod: function () { ... },
  } );

  Object.assign( Points.prototype, {
    deprecatedMethod: function () { ... },
  } );

}
```

**app.js**

```js
import { Mesh, Points, DeprecatedClass1, useDeprecatedMethods } from 'three';

enableDeprecatedMethods();

// ...
```

**Comment by marcofugaro** [comment]
> I was imagining just reorganizing src/Three.legacy.js, and leaving it in build/three.module.js, with sideEffects: false set. 

Let me test if this would work.

**Comment by marcofugaro** [comment]
@donmccurdy I tested your solution using webpack and rollup, turns out it works! 🎉 

The `sideEffects: false`  flag is not relevant at all in this scenario. It's used for the code that pollutes the global scope such as `polyfills.js` which injects the `Number` class. Polyfills are not included in `three.module.js` so it doesn't apply.

Also I noticed that webpack tree-shakes three by default in recent versions, just like rollup has been doing. This makes the `sideEffects` flag not necessary anymore 🎉

-------------

# Test results

## Current situation (dev)
```js
import * as THREE from 'three'
const renderer = new THREE.WebGLRenderer()
console.log(renderer)
```
This is the resulting minified bundle size:

<img src="https://user-images.githubusercontent.com/7217420/107627498-ade39c00-6c5f-11eb-83e4-24f6294b6e8f.png" width="200" />


## Ideal situation

`Three.legacy.js` simply removed from bundle.

<img width="229" alt="legacy not included" src="https://user-images.githubusercontent.com/7217420/107627627-dd92a400-6c5f-11eb-92d3-a44839037502.png">

## Methods wrapped in function

Legacy methods are wrapped in the `enableDeprecatedMethods` function as illustrated by @donmccurdy.

The deprecated classes (`MeshFaceMaterial`, `MultiMaterial`), also are three-shaken, they are included only if the user explicitly calls them.

<img width="224" alt="legacy wrapped" src="https://user-images.githubusercontent.com/7217420/107627835-1cc0f500-6c60-11eb-9245-751e4e7e61b8.png">

## Methods wrapped in function and function called

```js
import * as THREE from 'three'
THREE.enableDeprecatedMethods()
const renderer = new THREE.WebGLRenderer()
console.log(renderer)
```

All the methods and respective classes are included, resulting in the initial bundle size.

<img width="239" alt="enableMethods called" src="https://user-images.githubusercontent.com/7217420/107628314-c2746400-6c60-11eb-8386-58c9aa833aca.png">

## Methods wrapped in function (rollup)

Situation is exactly the same with rollup

<img width="236" alt="Screenshot 2021-02-11 at 11 36 42" src="https://user-images.githubusercontent.com/7217420/107628854-8988bf00-6c61-11eb-8103-582d57575914.png">

## Methods wrapped in function and function called (rollup)
<img width="243" alt="Screenshot 2021-02-11 at 11 37 58" src="https://user-images.githubusercontent.com/7217420/107628863-8d1c4600-6c61-11eb-9b4a-c9e44338d259.png">

**Comment by ianpurvis** [comment]
Cool, glad it worked! Thanks for the test @marcofugaro

**Comment by donmccurdy** [comment]
Awesome, thanks @marcofugaro! 

The [Webpack 5 announcement](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#inner-module-tree-shaking) still mentions `sideEffects: false` as an important thing, so I'd be inclined to set it after we can do so safely, still.

**Comment by marcofugaro** [comment]
@donmccurdy I see, I guess this situation is happening, from the article you posted:

> webpack 5 can also automatically flag modules as side-effect-free according to a static analysis of the source code

Still, we can set it just to be sure, as you say.

**Comment by Filyus** [comment]
Looks like there not all files in the PRs (from Loaders for example). Can someone list all remaining files?

**Comment by Mugen87** [comment]
Like mentioned in https://github.com/mrdoob/three.js/pull/21292#issuecomment-779939203 I think we should completely get away from `Three.Legacy.js`. The idea is to:

- delete all entries which are older than two years
- move the remaining stuff into the specific classes
- keep these and new warning for only one year or a similar time frame.

Wrapping the legacy code into `enableDeprecatedMethods()` seems to good approach to me since it is too inconvenient. Warnings should be logged without asking the user to call a method. If not, users could also expect a wrapper around `console` to configure the log level in general.

**Comment by DefinitelyMaybe** [comment]
[comment from above](https://github.com/mrdoob/three.js/issues/19986#issuecomment-776161787)
re:
> * move the remaining stuff into the specific classes

unless I've misunderstood what was said

**Comment by Mugen87** [comment]
> Correct. It was horrible.

I don't think it would be that bad. I'm using this approach in other projects too and it works just fine.

**Comment by DefinitelyMaybe** [comment]
alrite, would it be within kooee to draft a small PR making the case?

**Comment by donmccurdy** [comment]
Moving deprecated methods into the relevant source files would prevent tree-shaking of the methods, but if we're OK with significantly reducing the _number_ of deprecated methods over time then that is fine. This suggestion requires feedback from @mrdoob (re: https://github.com/mrdoob/three.js/issues/19986#issuecomment-776161787), I don't think it's worth drafting a PR without his thoughts first.

**Comment by danny-laa** [comment]
Is there any roadmap for making three.js tree shakable?

**Comment by DefinitelyMaybe** [comment]
If you know you're way around tree-shaking with various tools then please have a look into the code-base and tell us what you think, otherwise the roadmap is more or less: Yes, we're working on it.

**Comment by danny-laa** [comment]
You may take a look at [yushijinhun/three-minifier (v0.3.0-alpha.0)](https://github.com/yushijinhun/three-minifier/tree/v0.3.x), which makes tree shaking works in my webpack project.

**Comment by mrdoob** [comment]
@danny-laa https://github.com/mrdoob/three.js/pull/21313

**Comment by gkjohnson** [comment]
I'm less up to date on this whole thread but I saw there was some issue with setting the [prototype.is* fields](https://github.com/mrdoob/three.js/pull/21293) and [treeshaking](https://github.com/webpack/webpack/issues/12716). It was [mentioned above](https://github.com/mrdoob/three.js/issues/19986#issuecomment-670993952) but is there a reason that using a getter isn't acceptable? I didn't see a lot of discussion around it but maybe I missed it.

```js
class Vector3 {
    get isVector3() { return true; }
    // ...
}
```

I assume it would be tree shakeable and should otherwise behave (nearly) the same as setting it on the prototype. It's not iterable and wouldn't create new memory on the object instance. The change in behavior would be that it's no longer directly settable on the instance but that seems like odd behavior to try to preserve considering three.js internals rely on that field being set. And if a user _really_ wants the change it they can use `Object.defineProperty` or override the getter in a derivative class.

Don't mean to repeat a conversation if this has already been addressed!

**Comment by Mugen87** [comment]
https://discord.com/channels/685241246557667386/685241247233081362/814913888205275136

**Comment by marcofugaro** [comment]
Update on the `Three.Legacy.js` discussion:

After a bit of digging, I found out that only the getters/setters defined with the `Object.defineProperty()` are non-tree-shakeable (`Object.assign()` calls were removed).

For example, the class `Skeleton` is not tree-shakeable because of these few lines:

https://github.com/mrdoob/three.js/blob/4cfbf39965c5826e334d5c8a30178a194e4c6bad/src/Three.Legacy.js#L954-L967

Anybody has any solution regarding this?

BTW since #21437 you guys can test the tree-shaking directly in the three.js repo.

**Comment by mrdoob** [comment]
We can definitely remove that `useVertexTexture` getter/setter 👍

**Comment by marcofugaro** [comment]
Great! I've made a list of deprecated methods preventing tree-shaking that we may consider removing (I've not included the common `BufferGeometry`, `Mesh`, `Light` classes):

- `Audio.prototype.startTime` is preventing `Audio` from being tree-shaken.
- `LOD.objects` is preventing `LOD` from being tree-shaken.
- `Uniform.dynamic` and `Uniform.onUpdate` are preventing `Uniform` from being tree-shaken.
- `MeshPhongMaterial.metal` is preventing `MeshPhongMaterial` from being tree-shaken.

These 4 alone save ~12kb from the minified three-shaken bundle.

Also in the future, we might consider removing these ones as well:

- `InstancedBufferGeometry.maxInstancedCount` is preventing `InstancedBufferGeometry` from being tree-shaken, however `InstancedBufferGeometry` can't be tree-shaken because it's not an es6 class yet.
- `Curve.__arcLengthDivisions` is preventing `Curve` from being tree-shaken, however `Curve` can't be tree-shaken because it's not an es6 class yet.
- `Raycaster.linePrecision` is preventing `Raycaster` from being tree-shaken, however `Raycaster` can't be tree-shaken because it's not an es6 class yet.
- `InterleavedBuffer.dynamic` is preventing `InterleavedBuffer` from being tree-shaken, however `InterleavedBuffer` can't be tree-shaken yet because it's not an es6 class yet.
- `MeshPhysicalMaterial.transparency` is preventing `MeshPhysicalMaterial` from being tree-shaken, however `MeshPhysicalMaterial` can't be tree-shaken because it's not an es6 class yet.

**Comment by mrdoob** [comment]
Awesome! Thanks for looking into that.

> - `Audio.startTime` is preventing `Audio` from being tree-shaken.
> - `LOD.objects` is preventing `LOD` from being tree-shaken.
> - `Uniform.dynamic` and `Uniform.onUpdate` are preventing `Uniform` from being tree-shaken.
> - `MeshPhongMaterial.metal` is preventing `MeshPhongMaterial` from being tree-shaken.

I think it's safe to remove all these 👍

**Comment by marcofugaro** [comment]
Hey, does anybody know what the blockers of this feature are? Is there a reason we don't convert the remaining part of the library to classes and move on with this? Is there something that I can do to help? 🙂

**Comment by Mugen87** [comment]
@marcofugaro We need to solve #20527 first.

**Comment by marcofugaro** [comment]
Alright, will see it I can make it work starting from that PR

**Comment by mrdoob** [comment]
We had quite a few changes in core this cycle so I was cautious and tried to not have too many things going on.
I'll try to get some progress on this in the next cycle.

**Comment by marcofugaro** [comment]
@Mugen87 it is unclear to me, do we still need to support IE11 for the `examples/js` folder?

**Comment by Mugen87** [comment]
AFAIK, yes. The idea was that #20527 would provide the foundation for ES6->ES5 transpilation if users have to support oldish browsers.

**Comment by gkjohnson** [comment]
@Mugen87 looking at the `build/three.js` and `build/three.min.js` files they both contain `class`, `const`, and `let` keywords meaning the core files already won't work in IE11. What's the reasoning for only supporting IE11 in the `examples/js` files?

**Comment by Mugen87** [comment]
The user can easily configure the Babel preset in order to produce build files that work in IE11 again.

It would be good to have the same for `examples/js`.

**Comment by DefinitelyMaybe** [comment]
> What's the reasoning for only supporting IE11 in the examples/js files?

so not _only_ supporting, instead supporting when needed by user, yes?

**Comment by mrdoob** [comment]
> @Mugen87 it is unclear to me, do we still need to support IE11 for the `examples/js` folder?

Last week's release (`r127`) was the first one to not have IE11 support in the builds.
I think we can go ahead and do the same for `examples/js` now.

**Comment by Mugen87** [comment]
When I remember correctly the modularization script can not handle classes, yet.

Wouldn't it be better to solve #20527 before investing more time in `utils/modularize.js`?

**Comment by marcofugaro** [comment]
@Mugen87 I did a PR for the `unmodularize` script, it can replace the `modularize` one 🙂 

It supports classes and can optionally transpile with babel.

**Comment by DefinitelyMaybe** [comment]
wooh! time to tackle the examples folder, good stuff everyone 👍

**Comment by DefinitelyMaybe** [comment]
quick shout out to @Mugen87 ! 🎉 💯 👍 
That was a ton from the `examples` folder done over the past couple of days. Great stuff.

**Comment by mrdoob** [comment]
> quick shout out to @Mugen87 ! 🎉 💯 👍 
> That was a ton from the `examples` folder done over the past couple of days. Great stuff.

Indeed! What an herculean effort!👏👏👏

**Comment by Mugen87** [comment]
The only files in the core which have to be converted to classes are `WebGLRenderer` and its related components (e.g. `WebGLAnimation`, `WebGLState` etc.). Looking at this code, it has a specific structure which is not easy to transform to a ES6 class. It would require some sort of rewrite. I guess I would vote not do this unless it is really required.

Question: Is it necessary for improving tree-shaking to convert the remaining classes in `src/renderers`?

Otherwise we could consider this issue as done since `examples/jsm/nodes` does not need to be converted to ES6 classes (assuming it is replaced with the new node system in `examples/jsm/renderers/nodes`).

**Comment by drcmda** [comment]
btw, do you guys know about [lebab](https://lebab.unibtc.me/)? it's a tool that does the conversion from old to new (babel backwards) almost automatically. for three-stdlib we did that: classes, templates, const, let, =>, object spread, object shorthand, etc. the whole examples folder was done in a few hours, with a few things left to patch in between. if there's still some work left, maybe it can still help.

**Comment by Mugen87** [comment]
I've made the experience that `lebab` can not handle all class definitions into a ES6 class. Hence, it was more reliable to do this manually.

**Comment by marcofugaro** [comment]
> Question: Is it necessary for improving tree-shaking to convert the remaining classes in src/renderers?

`WebGLAnimation` is used by `WebXRManager`, which is used directly in the `WebGLRenderer` in this line here:

https://github.com/mrdoob/three.js/blob/be270aef9f0c5266c19100f0384f52928cbc8095/src/renderers/WebGLRenderer.js#L314

So unless the API changes to something similar to what discussed in  #19513, `WebXRManager` and `WebGLAnimation` won't get tree-shaken.


Same goes for `WebGLState` and other classes used directly in the `WebGLRender`, they won't get tree-shaken automatically if converted to classes because they're always in use.

**Comment by mrdoob** [comment]
@marcofugaro I wonder if `WebGLAnimation` and `WebXRManager` are worth the effort though...

Should we add a tree map to the treeshaking script to see if there are easier wins?

**Comment by marcofugaro** [comment]
> Should we add a tree map (rollup-plugin-visualizer) to the treeshaking script to see if there are easier wins?

Sure! Will look into it!

By the way another easier win to remove are the deprecated methods of the root classes such as `WebGLRenderer`, `WebGLRenderTarget` and so on. They occupy `~30kb` of the minified bundle of the user.

They are not removed automatically [for this reason](https://rollupjs.org/repl/?version=2.45.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiolMjBNQUlOJTIwRklMRSUyMColMkYlNUNuaW1wb3J0JTIwJTdCJTIwVXNlZENsYXNzJTIwJTdEJTIwZnJvbSUyMCcuJTJGbWF0aHMuanMnJTNCJTVDbiU1Q25jb25zb2xlLmxvZyglMjBuZXclMjBVc2VkQ2xhc3MoKSUyMCklM0IlMjAlMkYlMkYlMjAxMjUlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJtYXRocy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjIlMkYlMkYlMjBtYXRocy5qcyU1Q24lNUNuZXhwb3J0JTIwY2xhc3MlMjBVc2VkQ2xhc3MlMjAlN0IlNUNuJTVDdGNvbnN0cnVjdG9yKHgpJTIwJTdCJTVDbiU1Q3QlNUN0JTIwdGhpcy5yZXN1bHQlMjAlM0QlMjB4JTIwKiUyMHglMjAqJTIweCUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEJTVDbiUyRiUyRiUyMHRoaXMlMjBpcyUyMG5vdCUyMHJlbW92ZWQlMjBldmVuJTIwaWYlMjBub3QlMjBjYWxsZWQlNUNuVXNlZENsYXNzLnByb3RvdHlwZS5kZXByZWNhdGVkTWV0aG9kJTIwJTNEJTIwZnVuY3Rpb24lMjAoKSUyMCU3QiUyMGNvbnNvbGUubG9nKCdERVBSRUNBVEVEJyklMjAlN0QlNUNuJTVDbmV4cG9ydCUyMGNsYXNzJTIwVW51c2VkQ2xhc3MlMjAlN0IlNUNuJTIwJTIwcHJvcCUyMCUzRCUyMDMlM0IlNUNuJTdEJTVDblVudXNlZENsYXNzLnByb3RvdHlwZS5kZXByZWNhdGVkTWV0aG9kJTIwJTNEJTIwZnVuY3Rpb24lMjAoKSUyMCU3QiUyMGNvbnNvbGUubG9nKCdERVBSRUNBVEVEJyklMjAlN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==), and I couldn't find a way to make them tree-shake. Does anybody have any input about this?

**Comment by Mugen87** [comment]
It seems we can consider this issue as done, right? AFAICS, the ES6 class conversion is finished since the remaining folders `src/renderer` and `examples/jsm/nodes` are not changed.

**Comment by mrdoob** [comment]
Sounds good to me 👍

Excellent work everyone! Many thanks for helping out! 🙏
